File: src/mycli/services/hooks/management.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from mycli.services.hooks.allowlist import HookAllowlist, command_digest
from mycli.services.hooks.config import (
    ConfiguredHookSpec,
    HookConfigDiscovery,
    HookConfigIssue,
    HookConfigRegistry,
)
from mycli.services.hooks.types import HookPoint


@dataclass(slots=True, frozen=True)
class HookManagementRow:
    source: str
    hook_id: str
    hook_point: str
    identity: str
    enabled: bool
    timeout_seconds: float
    working_directory: str
    env_policy: str
    command_digest: str
    allowlist_status: str
    allowlist_reason: str
    config_path: str
    config_issues: tuple[str, ...]
    allowlist_issues: tuple[str, ...]

    def to_dict(self) -> dict[str, object]:
        return {
            "source": self.source,
            "hook_id": self.hook_id,
            "hook_point": self.hook_point,
            "identity": self.identity,
            "enabled": self.enabled,
            "timeout_seconds": self.timeout_seconds,
            "working_directory": self.working_directory,
            "env_policy": self.env_policy,
            "command_digest": self.command_digest,
            "allowlist_status": self.allowlist_status,
            "allowlist_reason": self.allowlist_reason,
            "config_path": self.config_path,
            "config_issues": list(self.config_issues),
            "allowlist_issues": list(self.allowlist_issues),
        }


@dataclass(slots=True, frozen=True)
class HookManagementResponse:
    ok: bool
    action: str
    message: str
    hooks: tuple[HookManagementRow, ...] = ()
    hook: HookManagementRow | None = None
    config_issues: tuple[str, ...] = ()
    allowlist_issues: tuple[str, ...] = ()
    removed: bool | None = None

    def to_dict(self) -> dict[str, object]:
        payload: dict[str, object] = {
            "ok": self.ok,
            "action": self.action,
            "message": self.message,
            "hooks": [row.to_dict() for row in self.hooks],
            "config_issues": list(self.config_issues),
            "allowlist_issues": list(self.allowlist_issues),
        }
        if self.hook is not None:
            payload["hook"] = self.hook.to_dict()
        if self.removed is not None:
            payload["removed"] = self.removed
        return payload


class HookManagementService:
    def __init__(
        self,
        *,
        workspace_root: Path,
        home_dir: Path,
        shell_path: str | None = None,
    ) -> None:
        self._workspace_root = workspace_root
        self._home_dir = home_dir
        self._shell_path = shell_path

    def list_hooks(self) -> HookManagementResponse:
        discovery, allowlist = self._load()
        rows = self._rows(discovery.hooks, discovery.issues, allowlist)
        return HookManagementResponse(
            ok=True,
            action="list",
            message=f"configured hooks: {len(rows)}",
            hooks=rows,
            config_issues=_safe_config_issues(discovery.issues),
            allowlist_issues=allowlist.issues,
        )

    def inspect_hook(self, identity: str) -> HookManagementResponse:
        discovery, allowlist = self._load()
        spec = self._resolve(discovery.hooks, identity)
        if spec is None:
            return HookManagementResponse(
                ok=False,
                action="inspect",
                message=f"configured hook not found: {identity}",
                config_issues=_safe_config_issues(discovery.issues),
                allowlist_issues=allowlist.issues,
            )
        row = self._row(spec, discovery.issues, allowlist)
        return HookManagementResponse(
            ok=True,
            action="inspect",
            message=f"configured hook: {row.identity}",
            hook=row,
            hooks=(row,),
            config_issues=_safe_config_issues(discovery.issues),
            allowlist_issues=allowlist.issues,
        )

    def approve_hook(self, identity: str) -> HookManagementResponse:
        discovery, allowlist = self._load()
        spec = self._resolve(discovery.hooks, identity)
        if spec is None:
            return HookManagementResponse(
                ok=False,
                action="approve",
                message=f"configured hook not found: {identity}",
                config_issues=_safe_config_issues(discovery.issues),
                allowlist_issues=allowlist.issues,
            )
        if allowlist.issues:
            return HookManagementResponse(
                ok=False,
                action="approve",
                message="allowlist cannot be written until parse issues are fixed",
                hook=self._row(spec, discovery.issues, allowlist),
                config_issues=_safe_config_issues(discovery.issues),
                allowlist_issues=allowlist.issues,
            )
        try:
            allowlist.approve(spec)
        except (OSError, ValueError) as exc:
            return HookManagementResponse(
                ok=False,
                action="approve",
                message=f"allowlist write failed: {exc}",
                hook=self._row(spec, discovery.issues, allowlist),
                config_issues=_safe_config_issues(discovery.issues),
                allowlist_issues=allowlist.issues,
            )
        refreshed = HookAllowlist(home_dir=self._home_dir)
        row = self._row(spec, discovery.issues, refreshed)
        return HookManagementResponse(
            ok=True,
            action="approve",
            message=f"approved {row.identity}",
            hook=row,
            hooks=(row,),
            config_issues=_safe_config_issues(discovery.issues),
            allowlist_issues=refreshed.issues,
        )

    def revoke_hook(self, identity: str) -> HookManagementResponse:
        discovery, allowlist = self._load()
        spec = self._resolve(discovery.hooks, identity)
        if spec is None:
            return HookManagementResponse(
                ok=False,
                action="revoke",
                message=f"configured hook not found: {identity}",
                config_issues=_safe_config_issues(discovery.issues),
                allowlist_issues=allowlist.issues,
            )
        if allowlist.issues:
            return HookManagementResponse(
                ok=False,
                action="revoke",
                message="allowlist cannot be written until parse issues are fixed",
                hook=self._row(spec, discovery.issues, allowlist),
                config_issues=_safe_config_issues(discovery.issues),
                allowlist_issues=allowlist.issues,
            )
        try:
            removed = allowlist.revoke(spec)
        except (OSError, ValueError) as exc:
            return HookManagementResponse(
                ok=False,
                action="revoke",
                message=f"allowlist write failed: {exc}",
                hook=self._row(spec, discovery.issues, allowlist),
                config_issues=_safe_config_issues(discovery.issues),
                allowlist_issues=allowlist.issues,
            )
        refreshed = HookAllowlist(home_dir=self._home_dir)
        row = self._row(spec, discovery.issues, refreshed)
        return HookManagementResponse(
            ok=True,
            action="revoke",
            message=f"revoked {row.identity}" if removed else f"not approved {row.identity}",
            hook=row,
            hooks=(row,),
            config_issues=_safe_config_issues(discovery.issues),
            allowlist_issues=refreshed.issues,
            removed=removed,
        )

    def _load(self) -> tuple[HookConfigDiscovery, HookAllowlist]:
        discovery = HookConfigRegistry(
            workspace_root=self._workspace_root,
            home_dir=self._home_dir,
            shell_path=self._shell_path,
        ).discover()
        return discovery, HookAllowlist(home_dir=self._home_dir)

    def _rows(
        self,
        specs: tuple[ConfiguredHookSpec, ...],
        issues: tuple[HookConfigIssue, ...],
        allowlist: HookAllowlist,
    ) -> tuple[HookManagementRow, ...]:
        return tuple(self._row(spec, issues, allowlist) for spec in specs)

    def _row(
        self,
        spec: ConfiguredHookSpec,
        issues: tuple[HookConfigIssue, ...],
        allowlist: HookAllowlist,
    ) -> HookManagementRow:
        status = allowlist.status_for(spec)
        return HookManagementRow(
            source=spec.source.value,
            hook_id=spec.hook_id,
            hook_point=spec.hook_point.value,
            identity=hook_identity(spec),
            enabled=spec.enabled,
            timeout_seconds=spec.timeout_seconds,
            working_directory=spec.working_directory.value,
            env_policy=spec.env_policy.value,
            command_digest=command_digest(spec.command),
            allowlist_status="allowed" if status.allowed else "not_allowed",
            allowlist_reason=status.reason,
            config_path=str(spec.source_path or ""),
            config_issues=tuple(
                issue.safe_line()
                for issue in issues
                if spec.source_path is not None and issue.path == spec.source_path
            ),
            allowlist_issues=allowlist.issues,
        )

    def _resolve(
        self,
        specs: tuple[ConfiguredHookSpec, ...],
        identity: str,
    ) -> ConfiguredHookSpec | None:
        parsed = parse_hook_identity(identity)
        if parsed is None:
            return None
        source, hook_id, hook_point = parsed
        for spec in specs:
            if (
                spec.source.value == source
                and spec.hook_id == hook_id
                and spec.hook_point is hook_point
            ):
                return spec
        return None


def hook_identity(spec: ConfiguredHookSpec) -> str:
    return f"{spec.source.value}:{spec.hook_id}:{spec.hook_point.value}"


def parse_hook_identity(identity: str) -> tuple[str, str, HookPoint] | None:
    parts = identity.split(":")
    if len(parts) != 3:
        return None
    source, hook_id, hook_point_value = (part.strip() for part in parts)
    if source not in {"repo", "user"} or not hook_id:
        return None
    try:
        hook_point = HookPoint(hook_point_value)
    except ValueError:
        return None
    return source, hook_id, hook_point


def _safe_config_issues(issues: tuple[HookConfigIssue, ...]) -> tuple[str, ...]:
    return tuple(issue.safe_line() for issue in issues)

File: backend/packages/integrations/src/hooks/management.ts
import { HookAllowlistStore, hookCommandDigest, hookConfigPathHash, hookIdentity } from "./allowlist.ts";
import { discoverHookConfig } from "./config.ts";
import type { DiscoverHookConfigOptions } from "./config.ts";
import type {
	ConfiguredHookSpec,
	HookApprovalStatus,
	HookConfigDiagnostic,
} from "./types.ts";

export interface HookManagementRow {
	readonly scope: ConfiguredHookSpec["scope"];
	readonly hookId: string;
	readonly hookPoint: ConfiguredHookSpec["hookPoint"];
	readonly identity: string;
	readonly enabled: boolean;
	readonly timeoutMs: number;
	readonly workingDirectory: ConfiguredHookSpec["workingDirectory"];
	readonly envPolicy: ConfiguredHookSpec["envPolicy"];
	readonly shellKind?: ConfiguredHookSpec["shellKind"];
	readonly commandDigest: string;
	readonly configPathHash: string;
	readonly allowlistStatus: "allowed" | "not_allowed";
	readonly allowlistReason: HookApprovalStatus["reason"];
}

export interface HookManagementResponse {
	readonly ok: boolean;
	readonly action: "list" | "inspect" | "approve" | "revoke";
	readonly message: string;
	readonly hooks: readonly HookManagementRow[];
	readonly hook?: HookManagementRow;
	readonly issues: readonly string[];
	readonly removed?: boolean;
}

export interface HookManagementServiceOptions extends DiscoverHookConfigOptions {
	readonly allowlistStore?: HookAllowlistStore;
}

export class HookManagementService {
	readonly #options: HookManagementServiceOptions;
	readonly #allowlist: HookAllowlistStore;

	constructor(options: HookManagementServiceOptions) {
		this.#options = options;
		this.#allowlist = options.allowlistStore ?? new HookAllowlistStore({ homeDir: options.homeDir });
	}

	async list(): Promise<HookManagementResponse> {
		const state = await this.#load();
		return response({
			ok: true,
			action: "list",
			message: `configured hooks: ${state.rows.length}`,
			hooks: state.rows,
			issues: state.issues,
		});
	}

	async inspect(identity: string): Promise<HookManagementResponse> {
		const state = await this.#load();
		const hook = state.rows.find((row) => row.identity === identity && validIdentity(identity));
		return response({
			ok: hook !== undefined,
			action: "inspect",
			message: hook
				? `configured hook: ${hook.identity}`
				: `configured hook not found: ${safeIdentity(identity)}`,
			hooks: hook ? [hook] : [],
			...(hook ? { hook } : {}),
			issues: state.issues,
		});
	}

	async approve(identity: string): Promise<HookManagementResponse> {
		return this.#mutate("approve", identity);
	}

	async revoke(identity: string): Promise<HookManagementResponse> {
		return this.#mutate("revoke", identity);
	}

	async #mutate(
		action: "approve" | "revoke",
		identity: string,
	): Promise<HookManagementResponse> {
		const discovery = await this.#discover();
		const spec = discovery.hooks.find((hook) => (
			validIdentity(identity) && hookIdentity(hook) === identity
		));
		const initialIssues = await this.#issues(discovery.diagnostics);
		if (!spec) {
			return response({
				ok: false,
				action,
				message: `configured hook not found: ${safeIdentity(identity)}`,
				hooks: [],
				issues: initialIssues,
			});
		}
		let removed: boolean | undefined;
		try {
			if (action === "approve") await this.#allowlist.approve(spec);
			else removed = await this.#allowlist.revoke(spec);
		} catch {
			return response({
				ok: false,
				action,
				message: "configured hook allowlist update failed",
				hooks: [],
				issues: await this.#issues(discovery.diagnostics),
			});
		}
		const hook = await managementRow(spec, this.#allowlist);
		return response({
			ok: true,
			action,
			message: action === "approve"
				? `approved ${hook.identity}`
				: removed
					? `revoked ${hook.identity}`
					: `not approved ${hook.identity}`,
			hooks: [hook],
			hook,
			issues: await this.#issues(discovery.diagnostics),
			...(removed === undefined ? {} : { removed }),
		});
	}

	async #load(): Promise<{
		readonly rows: readonly HookManagementRow[];
		readonly issues: readonly string[];
	}> {
		const discovery = await this.#discover();
		const rows = await Promise.all(discovery.hooks.map((hook) => (
			managementRow(hook, this.#allowlist)
		)));
		return Object.freeze({
			rows: Object.freeze(rows),
			issues: await this.#issues(discovery.diagnostics),
		});
	}

	#discover() {
		return discoverHookConfig({
			workspaceRoot: this.#options.workspaceRoot,
			homeDir: this.#options.homeDir,
			...(this.#options.includeRepository === undefined
				? {}
				: { includeRepository: this.#options.includeRepository }),
			...(this.#options.platform ? { platform: this.#options.platform } : {}),
			...(this.#options.env ? { env: this.#options.env } : {}),
			...(this.#options.shellPath ? { shellPath: this.#options.shellPath } : {}),
		});
	}

	async #issues(configDiagnostics: readonly HookConfigDiagnostic[]): Promise<readonly string[]> {
		const configIssues = configDiagnostics.map((item) => (
			`${item.scope}:${item.fileLabel}:${item.hookId}:${item.errorClass}`
		));
		const allowlist = await this.#allowlist.load();
		return Object.freeze([
			...configIssues,
			...allowlist.issues.map((issue) => `allowlist:${issue}`),
		]);
	}
}

async function managementRow(
	spec: ConfiguredHookSpec,
	allowlist: Pick<HookAllowlistStore, "statusFor">,
): Promise<HookManagementRow> {
	const approval = await allowlist.statusFor(spec);
	return Object.freeze({
		scope: spec.scope,
		hookId: spec.hookId,
		hookPoint: spec.hookPoint,
		identity: hookIdentity(spec),
		enabled: spec.enabled,
		timeoutMs: spec.timeoutMs,
		workingDirectory: spec.workingDirectory,
		envPolicy: spec.envPolicy,
		...(spec.shellKind ? { shellKind: spec.shellKind } : {}),
		commandDigest: hookCommandDigest(spec.command),
		configPathHash: hookConfigPathHash(spec.configPath),
		allowlistStatus: approval.allowed ? "allowed" : "not_allowed",
		allowlistReason: approval.reason,
	});
}

function response(input: HookManagementResponse): HookManagementResponse {
	return Object.freeze({
		...input,
		hooks: Object.freeze([...input.hooks]),
		issues: Object.freeze([...input.issues]),
	});
}

function validIdentity(value: string): boolean {
	return /^(?:user|repo):[A-Za-z0-9][A-Za-z0-9._-]{0,63}:(?:pre_tool_use|post_tool_use|user_prompt_submit|stop|pre_compact|session_start|session_end)$/u.test(value);
}

function safeIdentity(value: string): string {
	return validIdentity(value) ? value : "hook";
}
